cmd/cryptography/checksum.go

```diff
@@ -0,0 +1,176 @@
+package cryptography
+
+import (
+	"archive/zip"
+	"bufio"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	errs "github.com/open-cmsis-pack/cpackget/cmd/errors"
+	"github.com/open-cmsis-pack/cpackget/cmd/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+// isValidHash returns whether a hash function is
+// supported or not.
+func isValidHash(hashFunction string) bool {
+	for _, h := range Hashes {
+		if h == hashFunction {
+			return true
+		}
+	}
+	return false
+}
+
+// getChecksumList computes the digests of a pack according
+// to the specified hash function.
+func getChecksumList(sourcePack, hashFunction string) (map[string]string, error) {
+	var h hash.Hash
+	switch hashFunction {
+	case "sha256":
+		h = sha256.New()
+	} // Default will always be "sha256" if nothing is passed
+
+	zipReader, err := zip.OpenReader(sourcePack)
+	if err != nil {
+		log.Errorf("can't decompress \"%s\": %s", sourcePack, err)
+		return nil, errs.ErrFailedDecompressingFile
+	}
+
+	digests := make(map[string]string)
+	for _, file := range zipReader.File {
+		reader, err := file.Open()
+		if err != nil {
+			return nil, err
+		}
+		// Avoid Potential DoS vulnerability via decompression bomb
```

Comment:
Awesome security check!

It's just missing one check tho: you need to set a hard limit for the decompressed file, otherwise the shell bomb might still happen, even if you're copying 1KB at a time and checking for EOF (which is what a shell bomb zip file will never provide you :) )

You can re-use this one: https://github.com/Open-CMSIS-Pack/cpackget/blob/main/cmd/utils/security.go#L47

Which I set the hard limit of 20G for a file. I don't imagine a zip in a file to be bigger than that.

Reply:
Done! Best to just use SecureCopy then!